packages/edit-site/src/hooks/use-theme-style-variations/use-theme-style-variations-by-property.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { mergeBaseAndUserConfigs } from '../../components/global-styles/global-styles-provider';
+import cloneDeep from '../../utils/clone-deep';
+
+/**
+ * Returns a new object with only the properties specified in `properties`.
+ *
+ * @param {Object} object   The object to filter
+ * @param {Object} property The property to filter by
+ * @return {Object} The merged object.
+ */
+export const filterObjectByProperty = ( object, property ) => {
+	if ( ! object ) {
+		return {};
+	}
+
+	const newObject = {};
+	Object.keys( object ).forEach( ( key ) => {
+		if ( key === property ) {
+			newObject[ key ] = object[ key ];
+		} else if ( typeof object[ key ] === 'object' ) {
+			const newFilter = filterObjectByProperty( object[ key ], property );
+			if ( Object.keys( newFilter ).length ) {
+				newObject[ key ] = newFilter;
+			}
+		}
+	} );
+	return newObject;
+};
+
+/**
+ * Returns a new object with only the properties specified in `properties`.
+ *
+ * @param {Object}   props               Object of hook args.
+ * @param {Object[]} props.variations    The theme style variations to filter.
+ * @param {string}   props.property      The property to filter by.
+ * @param {Function} props.filter        Optional. The filter function to apply to the variations.
+ * @param {Object}   props.baseVariation Optional. Base or user settings to be updated with variation properties.
+ * @return {Object[]|*} The merged object.
+ */
+export default function useThemeStyleVariationsByProperty( {
+	variations,
+	property,
+	filter,
+	baseVariation,
+} ) {
+	return useMemo( () => {
+		if ( ! property || ! variations || variations?.length === 0 ) {
+			return variations;
+		}
+
+		let processedStyleVariations = variations.map( ( variation ) => ( {
+			...filterObjectByProperty( variation, property ),
+			// Add variation title and description to every variation item.
+			title: variation?.title,
+			description: variation?.description,
+		} ) );
+
+		if (
+			typeof baseVariation === 'object' &&
+			Object.keys( baseVariation ).length > 0
+		) {
+			/*
+			 * Overwrites all baseVariation object `styleProperty` properties
+			 * with the theme variation `styleProperty` properties.
+			 */
+			const clonedBaseVariation = cloneDeep( baseVariation );
+			processedStyleVariations = processedStyleVariations.map(
```

Comment:
Tiny nit: this iterates over `variations` a second time. Would it be worth combining the logic into the one `variations.map` callback?

Reply:
It would be worth it. 😄 Something like this do you mean?

```js
		let processedStyleVariations = variations.map( ( variation ) => {
			let result = {
				...filterObjectByProperty( variation, property ),
				// Add variation title and description to every variation item.
				title: variation?.title,
				description: variation?.description,
			};

			if ( hasBaseVariation ) {
				/*
				 * Overwrites all baseVariation object `styleProperty` properties
				 * with the theme variation `styleProperty` properties.
				 */
				result = mergeBaseAndUserConfigs( clonedBaseVariation, result );
			}
			return result;
		} );
```

Reply:
Yep, that's exactly what I had in mind!